plugins/sentinelone/komand_sentinelone/tasks/monitor_activities_and_events/task.py

```diff
@@ -0,0 +1,327 @@
+
+    def log_pagination_cycle(self, cursors) -> None:
+        """
+        Provide logging information if task is in pagination cycle
+        """
+        pagination_info_string = "Pagination tokens received... "
+        for log_type, cursor in cursors.items():
+            if cursor is not None:
+                pagination_info_string += f" {log_type}:{cursor}"
+        self.logger.info("Paginating cycle in progress.")
+        self.logger.info(pagination_info_string)
+
+    def collect_logs_handler(
+        self,
+        queries: Dict,
+        cursors: Dict,
+        state: Dict,
+        lookback_timestamp: str,
+        current_run_timestamp: str,
+        total_forbidden_responses: int,
+        is_paginating: bool,
+    ) -> Tuple[List[Dict], int]:
+        """
+        Handle whether to query each endpoint based on input and pagination, gather returned data and error responses
+        """
+        all_logs = []
+        for log_type, check in queries.items():
+            if check:
+                cursor = cursors.get(log_type)
+                if is_paginating is False or (is_paginating is True and cursor):
+                    self.logger.info(f"Getting {log_type} between {lookback_timestamp} and {current_run_timestamp}...")
+                    if cursor:
+                        self.logger.info(f"Using next page cursor: {cursor}")
+                    logs, total_forbidden_responses = self.get_generic_logs(
+                        log_type,
+                        state,
+                        total_forbidden_responses,
+                        current_run_timestamp,
+                        lookback_timestamp,
+                        cursor,
+                    )
+                    self.logger.info(f"{len(logs)} logs found in {log_type} query")
+                    all_logs.extend(logs)
+                else:
+                    self.logger.info(f"Pagination in progress, skipping {log_type} collection...")
+        return all_logs, total_forbidden_responses
+
+    def get_generic_logs(
+        self,
+        log_type: str,
+        state: dict,
+        total_forbidden_responses: int,
+        current_run_timestamp: str,
+        last_run_timestamp: str = None,
+        pagination_cursor: str = None,
+    ) -> List[Dict]:
+        """
+        Determine log type and query that log set with query parameters determined by type, cursor, and last timestamp.
+        Set latest log timestamp and page cursor in state.
+        Return logs from query
+        """
+        timestamp_key = LOG_TYPE_MAP.get(log_type, {}).get("LAST_LOG_TIMESTAMP")
+        page_token_key = LOG_TYPE_MAP.get(log_type, {}).get("PAGE_TOKEN")
+        last_log_timestamp = state.get(timestamp_key)
+        if not last_log_timestamp:
+            last_log_timestamp = STARTING_TIMESTAMP
+        query_params = self.get_query_params(log_type, last_run_timestamp, current_run_timestamp, pagination_cursor)
```

Comment:
Should this last_run_timestamp not be the `last_log_timestamp` ? If the task runs at 10.15, but the last event log is 10.10 we should 10.10 in the next query?

Reply:
Hm I think it can still be last_run_timestamp. If the first query searches for everything between 10:05 to 10:10, the second query will search for everything between 10:10 and 10:15.

If the latest log timestamp from the first query is 10:07, then that is just the last event log up until 10:10 so nothing should exist between 10:07 and 10:10.